pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -0,0 +1,341 @@
+	// belong to the configured network so simply return it
+	if len(d.networkForRegistry) != 0 {
+		return d.networkForRegistry
+	}
+
+	// Try to determine the network by checking whether the endpoint IP belongs
+	// to any of the configure networks' CIDR ranges
+	if d.ranger == nil {
+		return ""
+	}
+	entries, err := d.ranger.ContainingNetworks(net.ParseIP(endpointIP))
+	if err != nil {
+		log.Errora(err)
+		return ""
+	}
+	if len(entries) == 0 {
+		return ""
+	}
+	if len(entries) > 1 {
+		log.Warnf("Found multiple networks CIDRs matching the endpoint IP: %s. Using the first match.", endpointIP)
+	}
+
+	return (entries[0].(namedRangerEntry)).name
+}
+
+// RegisterNotReadyEndpoints registers newly received NotReadyEndpoints
+// via MCP annotations
+func (d *MCPDiscovery) RegisterNotReadyEndpoints(conf *model.Config) {
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+
+	if nrEps, ok := conf.Annotations[notReadyEndpointkey]; ok {
+		addrs := strings.Split(nrEps, ",")
+		for _, addr := range addrs {
+			d.notReadyEndpoints[addr] = conf
+		}
+	}
+}
+
+func (d *MCPDiscovery) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
+	out := make([]*model.ServiceInstance, 0)
+
+	// There is only one IP for kube registry
+	proxyIP := proxy.IPAddresses[0]
+
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+	for addr, conf := range d.notReadyEndpoints {
+		notReadyIP, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			continue
+		}
+		notReadyPort, err := strconv.Atoi(port)
+		if proxyIP != notReadyIP || err != nil {
+			continue
+		}
+		endpoints := make([]*model.IstioEndpoint, 0)
+		hostname := kube.ServiceHostname(conf.Name, conf.Namespace, d.options.DomainSuffix)
+		se, ok := conf.Spec.(*networking.ServiceEntry)
+		if !ok {
+			return nil, errors.New("getProxyServiceInstances: wrong type")
+		}
+
+		for _, h := range se.Hosts {
+			for _, svcPort := range se.Ports {
+				// add not ready endpoint
+				out = append(out, notReadyServiceInstance(se, svcPort, conf, h, notReadyIP, notReadyPort))
+				// add other endpoints
+				out = append(out, serviceInstances(se, svcPort, conf, h)...)
+				var uid string
+				if d.options.mixerEnabled() {
+					uid = fmt.Sprintf("kubernetes://%s.%s", conf.Name, conf.Namespace)
+				}
+				// create an EDSEndpoint for notReady IP:Port
+				edsEndpoint := &model.IstioEndpoint{
+					Address:         notReadyIP,
+					EndpointPort:    uint32(notReadyPort),
+					ServicePortName: svcPort.Name,
+					Labels:          conf.Labels,
+					UID:             uid,
+					//	ServiceAccount:  ???, //TODO: nino-k
+					Network: d.endpointNetwork(notReadyIP),
+					//Locality:   ???, //TODO: nino-k
+					Attributes: model.ServiceAttributes{Name: conf.Name, Namespace: conf.Namespace},
+				}
+				endpoints = append(endpoints, edsEndpoint)
+			}
+			_ = d.options.XDSUpdater.EDSUpdate(d.options.ClusterID, string(hostname), conf.Namespace, endpoints)
```

Comment:
Why are we triggering an EDS update on `GetProxyServiceInstances`? Shouldn't this just be retrieving the instances?

Reply:
The edsUpdate is triggered to send the notReadyEndpoints, I guess `GetProxyServiceInstances` does two things, one is to get all the instances associated to a service the other is send an edsUpdate for the notReadyEndpoints

Reply:
I don't think we should overload this function, `GetProxyServiceInstances` is idempotent, I would expect I could call it with no side effects? Imagine we call this from `eds.go`, wouldn't that trigger an infinite loop?

Also, maybe I am missing context here, but why are we sending the not ready endpoints over EDS?

Reply:
@howardjohn I thought we were doing that in kube/controller, no? I'm trying to make this function to behave as close as possible in terms of handling NotReadyEndpoints to `GetProxyServiceInstances` in kube/controller. But maybe I need to verify that again since the logic seem to have changed slightly in there 🤔

Reply:
From a quick look, seems like we only call it in  `updateEDS` which is called in the InstanceHandler only, not GetProxyServiceInstances